### ISSUE TITLE
Youtube live streaming capture from NASA HD Earth Views seems very poor quality.  

### ISSUE THREAD

**Issue by liamkennedy** [streamlink/streamlink#5303]
### Checklist

- [X] This is a plugin issue and not a different kind of issue
- [X] [I have read the contribution guidelines](https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#contributing-to-streamlink)
- [X] [I have checked the list of open and recently closed plugin issues](https://github.com/streamlink/streamlink/issues?q=is%3Aissue+label%3A%22plugin+issue%22)
- [X] [I have checked the commit log of the master branch](https://github.com/streamlink/streamlink/commits/master)

### Streamlink version

Latest stable release

### Description

TESTED with version 5.3.1 and back to 2.2.0

NASA recently (Mar 23) switched all their live streaming from Ustream over to Youtube.  (Public NASA Ustream is now unavailable - even for NASA TV)

I have a particular interest in the HD live views of the earth - and the quality I am seeing captured using streamlink is very poor (remarkably bad in comparison to the old Ustream channel)

I'm not sure if this issue is related to the Youtube side of things - or if there is something I am doing wrong with streamlink. 

This is the command I am using:

`streamlink http://www.youtube.com/watch?v=itdpuGHAcpg best -o EHDC6-NASA.mp4`

The resulting capture is much lower quality than I would expect - showing significant blurriness between apparent "keyframes" (clear keyframes appear every 6 seconds or so). 

Here is an example capture https://youtu.be/vN9tkoqt2Pk 

The desktop stream (viewable in a browser directly from YouTube) looks "reasonable" - much higher quality than is captured in streamlink (but my judgment is the YouTube encoding is not as good as we used to see from Ustream - but that's unrelated to this specific issue I am reporting).  

Below is the debug output so you can see what's going on.  I don't see any errors, or something that jumps out that would explain the drop in quality. 

If anyone can confirm what I am seeing - or provide any guidance on how to improve that I will be very grateful. Thank you!

### Debug log

```text
G:\EHDC\2023-04-19>C:\Users\liamk\Downloads\streamlink-mobile\streamlink-5.3.1-1-py311-x86_64\bin\streamlink.exe http://www.youtube.com/watch?v=itdpuGHAcpg best -o EHDC6-NASA.mp4 -l debug
[cli][debug] OS:         Windows 10
[cli][debug] Python:     3.11.2
[cli][debug] Streamlink: 5.3.1
[cli][debug] Dependencies:
[cli][debug]  certifi: 2022.12.7
[cli][debug]  isodate: 0.6.1
[cli][debug]  lxml: 4.9.2
[cli][debug]  pycountry: 22.3.5
[cli][debug]  pycryptodome: 3.17
[cli][debug]  PySocks: 1.7.1
[cli][debug]  requests: 2.28.2
[cli][debug]  urllib3: 1.26.14
[cli][debug]  websocket-client: 1.5.1
[cli][debug] Arguments:
[cli][debug]  url=http://www.youtube.com/watch?v=itdpuGHAcpg
[cli][debug]  stream=['best']
[cli][debug]  --loglevel=debug
[cli][debug]  --output=EHDC6-NASA.mp4
[cli][debug]  --ffmpeg-ffmpeg=C:\Program Files (x86)\Streamlink\ffmpeg\ffmpeg.exe
[cli][info] Found matching plugin youtube for URL http://www.youtube.com/watch?v=itdpuGHAcpg
[plugins.youtube][debug] Using video ID: itdpuGHAcpg
[plugins.youtube][debug] This video is live.
[utils.l10n][debug] Language code: en_US
[cli][info] Available streams: 144p (worst), 240p, 360p, 480p, 720p (best)
[cli][info] Opening stream: 720p (hls)
[cli][info] Writing output to
G:\EHDC\2023-04-19\EHDC6-NASA.mp4
[cli][debug] Checking file output
[stream.hls][debug] Reloading playlist
[cli][debug] Pre-buffering 8192 bytes
[stream.hls][debug] First Sequence: 386678; Last Sequence: 386684
[stream.hls][debug] Start offset: 0; Duration: None; Start Sequence: 386682; End Sequence: None
[stream.hls][debug] Adding segment 386682 to queue
[stream.hls][debug] Adding segment 386683 to queue
[stream.hls][debug] Adding segment 386684 to queue
[stream.hls][debug] Writing segment 386682 to output
[stream.hls][debug] Segment 386682 complete
[cli][debug] Writing stream to output
[stream.hls][debug] Writing segment 386683 to output
[stream.hls][debug] Segment 386683 complete
[stream.hls][debug] Writing segment 386684 to output
[stream.hls][debug] Segment 386684 complete
[download] Written 6.75 MiB to EHDC6-NASA.mp4 (4s @ 1.54 MiB/s)                                                        [stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 386685 to queue
[stream.hls][debug] Adding segment 386686 to queue
[download] Written 6.75 MiB to EHDC6-NASA.mp4 (4s @ 1.45 MiB/s)                                                        [stream.hls][debug] Writing segment 386685 to output
[stream.hls][debug] Segment 386685 complete
[stream.hls][debug] Writing segment 386686 to output
[stream.hls][debug] Segment 386686 complete
[download] Written 11.28 MiB to EHDC6-NASA.mp4 (9s @ 1.17 MiB/s)
```

**Comment by bastimeyer** [comment]
Duplicate of #2936

Streamlink's YouTube plugin does only support HLS live streams. Any other implementation which YouTube's web player uses, like DASH for example, is not supported.

The stream you're seeing in your web browser uses the vp9 video codec, whereas the HLS stream is h264, potentially with a lower bitrate.

yt-dlp also only supports HLS (`yt-dlp -F 'https://www.youtube.com/watch?v=itdpuGHAcpg'`)

> `streamlink --stream-url 'https://www.youtube.com/watch?v=itdpuGHAcpg' best`
> `yt-dlp -s -j 'https://www.youtube.com/watch?v=itdpuGHAcpg' | jq -r '.formats[-1].url'`

**Comment by liamkennedy** [comment]
> Duplicate of #2936

Sorry, I didn't realize this issue with quality is related to issue #2936 - thanks for the link

> 
> Streamlink's YouTube plugin does only support HLS live streams. Any other implementation which YouTube's web player uses, like DASH for example, is not supported.
> 
> The stream you're seeing in your web browser uses the vp9 video codec, whereas the HLS stream is h264, potentially with a lower bitrate.

Yes - that's as I expected - the desktop stream and the hls version are not at all equivalent. It's just so odd how significantly worse it is.  

I've compared other youtube live streams (non NASA related) and the quality seems more acceptable even with the same restrictions.   

I'm curious if something is being handled differently either at the source encoding side of things (at NASA JSC) or if the re-encoding at YouTube is just not working quite right.  

> 
> yt-dlp also only supports HLS (`yt-dlp -F 'https://www.youtube.com/watch?v=itdpuGHAcpg'`)
> 
> > `streamlink --stream-url 'https://www.youtube.com/watch?v=itdpuGHAcpg' best`
> > `yt-dlp -s -j 'https://www.youtube.com/watch?v=itdpuGHAcpg' | jq -r '.formats[-1].url'`

So, bottom line, there is nothing to be done here. 

Have any of you checked out just how bad this particular stream looks from streamlink? It's just awful. (not that I am blaming it on streamlink - I'm just shocked how bad it is)

**Comment by bastimeyer** [comment]
> So, bottom line, there is nothing to be done here.

Unfortunately, yes.

YouTube's adaptive streams require URL query string signatures that need to be calculated from embedded JS code, which is a ton of work and maintenance burden if implemented. Somebody would need to submit a PR which implements this similar to yt-dlp, pytube or related projects (JS parser with simplified runtime, or Python reimplementation, etc). I don't want to go down that rabbit hole and waste my time with that.

And DASH streams are also still unsupported. Recently I've refactored/reworked/fixed Streamlink's DASH implementation, but there's still a ton to implement in order to get YT streams working.

Both adaptive streams and DASH streams provide vp9 video streams.

> Have any of you checked out just how bad this particular stream looks

The h264 video stream is really noisy, which is very obvious on clouds when showing ISS footage. I haven't seen such a difference in stream quality on any YT live stream yet.

I believe this is caused by the HLS stream having a variable bitrate instead of a constant one. The metadata says 4Mbit/s, but that's not the case on the actual content. It's an h264 video stream with the main profile. The vp9 stream apparently has a 1.2Mbit/s bitrate.
